projects/js-packages/components/components/boost-score-graph/annotations-plugin.ts

```diff
@@ -0,0 +1,70 @@
+import uPlot from 'uplot';
+import { Annotation } from '.';
+
+import './style-annotation.scss';
+
+// eslint-disable-next-line jsdoc/require-returns
+/**
+ * Custom tooltips plugin for uPlot.
+ *
+ * @param {Annotation[]} annotations - The periods to display in the tooltip.
+ */
+export function annotationsPlugin( annotations: Annotation[] ) {
+	let containerEl, annotationsContainer;
+
+	/**
+	 * Initialize the plugin
+	 *
+	 * @param {uPlot} u - The uPlot instance.
+	 */
+	function init( u: uPlot ) {
+		containerEl = u.under;
+
+		annotationsContainer = document.createElement( 'div' );
+
+		annotationsContainer.classList.add( 'jb-graph-annotations' );
+
+		const annotationEl = document.createElement( 'div' );
+		annotationEl.classList.add( 'jb-graph-annotations__annotation' );
+
+		annotations.forEach( annotation => {
+			const lineEl = document.createElement( 'div' );
+			lineEl.classList.add( 'jb-graph-annotations__line' );
+			lineEl.addEventListener( 'mouseenter', () => {
+				annotationEl.innerHTML = annotation.text;
+				annotationEl.style.display = 'block';
+				annotationEl.style.left = u.valToPos( annotation.timestamp / 1000, 'x' ) + 'px';
+			} );
+			lineEl.addEventListener( 'mouseleave', () => {
+				annotationEl.style.display = 'none';
+			} );
```

Comment:
Side note - make sure the event listeners don't leak, especially when changing routes without reloading the page. You might need a `removeEventListener` too

Reply:
AFAIK, when we remove nodes from DOM, the even listeners also get removed. And, navigating between routes does remove and recreate the nodes, so we should always have new nodes with fresh set of event listeners. Am I missing something?

Reply:
> If the event listener is not removed when the component unmounts, the event listener, its callback function, and your component will all stay in memory, causing a memory leak.

If you're adding one via the default API (addEventListener), you should always remove it.

Reply:
In this case, it is not a React component. The plugin is just another uPlot plugin that doesn't care if it's inside React or Vanila JS. The destroy and re-creation of the graph is handled by uPlot's React which handles destroying the [chart and thus nodes](https://github.com/skalinichev/uplot-wrappers/blob/master/react/uplot-react.tsx#L39). The chart is destroyed during unmounting.